### ISSUE TITLE
Overflow with flags enum

### ISSUE THREAD

**Issue by r00ster91** [crystal-lang/crystal#7636]
```cr
@[Flags]
enum E
  A
  B
  C
  D
  E
  F
  G
  H
  I
  J
  K
  L
  M
  N
  O
  P
  Q
  R
  S
  T
  U
  V
  W
  X
  Y
  Z
  A2
  B2
  C2
  D2
  E2
  F2
  G2
end

E::F2.to_i # => -2147483648
E::G2.to_i # => Overflow! 0
```
Having too many members in a flags enum causes an overflow. In the next version this would raise an `OverflowError` and would result in a compiler bug error.

I'd say the integer type that is used for enums should be changed to `UInt128` and then the maximum flag enum member count should be limited to 128.
An overflow can also happen when a normal enum has more than `Int32::MAX` (or `UInt128::MAX` when the above is done) members. That's probably a bit unlikely to ever happen but still possible so there should be a limit too I think.

**Comment by bcardiff** [comment]
Related to #7268

**Comment by straight-shoota** [comment]
@r00ster91 Do you have any reasonable use case for a flag enum with > 64 values?

**Comment by r00ster91** [comment]
No I don't have any use case. This should be fixed though.

**Comment by bcardiff** [comment]
Note that `E::G2.to_i` will always overflow since the return type is `Int32`, despite the enum been declared as `Int64` for example.

**Comment by yxhuvud** [comment]
Potentially, could the size of the uint be chosen to either be the least that fit all the flags?

**Comment by straight-shoota** [comment]
This has been fixed in #7776.
